src/main/java/com/wepay/zktools/clustermgr/internal/ClusterManagerImpl.java

```diff
@@ -207,24 +207,37 @@ public void manage(ManagedClient client) throws ClusterManagerException {
     @Override
     public void manage(ManagedServer server) throws ClusterManagerException {
         synchronized (serverManagementLock) {
+            int serverId;
             synchronized (managedServers) {
                 if (running) {
                     if (managedServers.containsKey(server)) {
-                        throw new ClusterManagerException("server already managed");
+                        serverId = managedServers.get(server).serverId;
+                        logger.info("updating the existing server");
+                    } else {
+                        serverId = nextId(serverIdZNode);
                     }
                 } else {
                     throw new ClusterManagerException("already closed");
                 }
             }
 
-            int serverId = nextId(serverIdZNode);
             ServerDescriptor descriptor =
                 new ServerDescriptor(serverId, server.endpoint(), server.getPreferredPartitions());
 
             try {
                 synchronized (managedServers) {
+                    ZNode oldZnode = null;
+                    if (managedServers.get(server) != null) {
+                        oldZnode = managedServers.get(server).znode;
+                    }
+
                     ZNode znode = createServerZNode(descriptor);
                     managedServers.put(server, new ManagedServerInfo(descriptor.serverId, znode));
+
+                    // Removed Znode of old server descriptor
+                    if (oldZnode != null) {
+                        zkClient.delete(oldZnode);
+                    }
```

Comment:
This action, creation of the new znode and deletion of the old node, is not atomic. There is a time window that two znodes for the same server coexist. This may break some code that assumes uniqueness. Use `setData()` to update the server info when there is an old node.

Reply:
fixed.